File: frontend/lib/services/functions/NoteService.dart
import 'dart:convert';
import 'dart:developer';
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:frontend/services/functions/AuthService.dart';
import 'package:frontend/utils/constants.dart';
import 'package:http/http.dart' as http;

class NoteService {
  String idToken = AuthService.idToken!;

  Stream<QuerySnapshot> getNoteStream(String enrollmentId) {
    return FirebaseFirestore.instance
        .collection('enrollments')
        .doc(enrollmentId)
        .collection('notes')
        .snapshots();
  }

  Future<String> addNote(String enrollmentId, Map<String, dynamic> data) async {
    final url = Uri.parse('${AppConstants.ENROLLMENT_API}/$enrollmentId/notes');
    final response = await http.post(
      url,
      headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer $idToken',
        },
      body: jsonEncode({'enrollmentId': enrollmentId, 'data': data}),
    );

    if (response.statusCode == 200) {
      final responseData = jsonDecode(response.body);
      return responseData['noteId'];
    } else {
      throw Exception('Failed to add note: ${response.body}');
    }
  }

  Future<void> deleteNote(String enrollmentId, String noteId) async {
    final url =
        Uri.parse('${AppConstants.ENROLLMENT_API}/$enrollmentId/$noteId');
    final response = await http.delete(url, headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer $idToken',
        },);

    if (response.statusCode != 200) {
      throw Exception('Failed to delete note: ${response.body}');
    }
  }

  Future<void> updateNote(
      String enrollmentId, String noteId, Map<String, dynamic> data) async {
    final url =
        Uri.parse('${AppConstants.ENROLLMENT_API}/$enrollmentId/$noteId');
    final response = await http.put(
      url,
      headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer $idToken',
        },
      body: jsonEncode(data),
    );

    if (response.statusCode != 200) {
      throw Exception('Failed to update note: ${response.body}');
    }
  }

  Future<List<Map<String, dynamic>>> getAllNotesOfEnrollment(
      String enrollmentId) async {
    final url = Uri.parse('${AppConstants.ENROLLMENT_API}/$enrollmentId/notes');
    final response = await http.get(url, headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer $idToken',
        },);

    if (response.statusCode == 200) {
      final responseData = jsonDecode(response.body) as List;
      return responseData.map((note) => note as Map<String, dynamic>).toList();
    } else {
      throw Exception('Failed to retrieve notes: ${response.body}');
    }
  }
}

File: backend/src/app/service/NoteService.js
const { EnrollmentCollection } = require('./Collections');


const addNote = async (enrollmentId, data) => {
    try {
        const noteRef = EnrollmentCollection.doc(enrollmentId).collection('notes').doc();
        await noteRef.set(data);
        return noteRef.id;
    } catch (error) {
        throw new Error(`Error adding note: ${error.message}`);
    }
};

const deleteNote = async (enrollmentId, noteId) => {
    try {
        await EnrollmentCollection.doc(enrollmentId).collection('notes').doc(noteId).delete();
    } catch (error) {
        throw new Error(`Error deleting note: ${error.message}`);
    }
};

const updateNote = async (enrollmentId, noteId, data) => {
    try {
        await EnrollmentCollection.doc(enrollmentId).collection('notes').doc(noteId).update(data);
    } catch (error) {
        throw new Error(`Error updating note: ${error.message}`);
    }
};

const getAllNotesOfEnrollment = async (enrollmentId) => {
    try {
        const notesSnapshot = await EnrollmentCollection.doc(enrollmentId).collection('notes').get();
        const notes = notesSnapshot.docs.map(doc => ({ id: doc.id, ...doc.data() }));
        return notes;
    } catch (error) {
        throw new Error(`Error retrieving notes: ${error.message}`);
    }
};

module.exports = {
    addNote,
    deleteNote,
    updateNote,
    getAllNotesOfEnrollment
};